src/signMessage.c

```diff
@@ -167,22 +165,83 @@ void handle_sign_message_parse_message(volatile unsigned int *tx) {
     }
 }
 
+static bool check_swap_validity(const SummaryItemKind_t kinds[MAX_TRANSACTION_SUMMARY_ITEMS],
+                                size_t num_summary_steps) {
+    bool amount_ok = false;
+    bool recipient_ok = false;
+    if (num_summary_steps != 2) {
+        PRINTF("2 steps expected for transaction in swap context, not %u\n", num_summary_steps);
+        return false;
+    }
+    for (size_t i = 0; i < num_summary_steps; ++i) {
+        transaction_summary_display_item(i, DisplayFlagNone | DisplayFlagLongPubkeys);
+        PRINTF("G_transaction_summary_title = %s\n", G_transaction_summary_title);
+        PRINTF("G_transaction_summary_text = %s\n", G_transaction_summary_text);
+        switch (kinds[i]) {
+            case SummaryItemAmount:
+                if (strcmp(G_transaction_summary_title, "Transfer") == 0) {
+                    amount_ok = check_swap_amount(G_transaction_summary_text);
+                } else {
+                    PRINTF("Refused field '%s', expecting 'Transfer'\n",
+                           G_transaction_summary_title);
+                    return false;
+                }
+                break;
+            case SummaryItemPubkey:
+                if (strcmp(G_transaction_summary_title, "Recipient") == 0) {
+                    recipient_ok = check_swap_recipient(G_transaction_summary_text);
+                } else {
+                    PRINTF("Refused field '%s', expecting 'Recipient'\n",
+                           G_transaction_summary_title);
+                    return false;
+                }
+                break;
+            default:
+                PRINTF("Refused kind '%u'\n", kinds[i]);
+                return false;
+        }
+    }
+    return amount_ok && recipient_ok;
+}
+
 void handle_sign_message_ui(volatile unsigned int *flags) {
     // Display the transaction summary
     SummaryItemKind_t summary_step_kinds[MAX_TRANSACTION_SUMMARY_ITEMS];
     size_t num_summary_steps = 0;
     if (transaction_summary_finalize(summary_step_kinds, &num_summary_steps) == 0) {
-        size_t num_flow_steps = 0;
+        // If we are in swap context, do not redisplay the message data
+        // Instead, ensure they are identitical with what was previously displayed
+        if (G_called_from_swap) {
+            if (G_swap_response_ready) {
```

Comment:
Is this scenario ever going to happen ? The app is supposed to exit after the the transaction is signed so... @fbeutin-ledger. Seems a bit unnecessary.

Reply:
The app is supposed to exit at the next loop after the `io_exchange()` call, currently there is no flow that makes this code necessary
However I'm wary of the app changing over time, and this code is an explicit protection against a malevolent client asking the app to make multiple payments.

Reply:
Ok. No worries. Better safe than sorry. Maybe add a comment saying this code is not supposed to be called because we should exit automatically ?

Reply:
Ok